search/search_repository.go

```diff
@@ -0,0 +1,185 @@
+	"golang.org/x/net/context"
+
+	"log"
+	"strconv"
+
+	"strings"
+
+	"regexp"
+
+	"github.com/almighty/almighty-core/app"
+	"github.com/almighty/almighty-core/models"
+	"github.com/asaskevich/govalidator"
+)
+
+const (
+	/*
+		- The SQL queries do a case-insensitive search.
+		- English words are normalized during search which means words like qualifying === qualify
+		- To disable the above normalization change "to_tsquery('english',$1)" to "to_tsquery($1)"
+		- Create GIN indexes : https://www.postgresql.org/docs/9.5/static/textsearch-tables.html#TEXTSEARCH-TABLES-INDEX
+		- To perform "LIKE" query we are appending ":*" to the search token
+
+	*/
+
+	// This SQL query is used when search is performed across workitem fields and workitem ID
+	testText = `select * from work_items WHERE
+		setweight(to_tsvector('english', id::text), 'A') ||
+		setweight(to_tsvector('english', fields::text), 'B') @@ to_tsquery('english',$1) and deleted_at is NULL`
+
+	// This SQL query is used when search is performed across workitem ID ONLY.
+	testID = `select * from work_items WHERE to_tsvector('english', id::text || ' ') @@ to_tsquery('english',$1) and deleted_at is NULL`
+)
+
+// GormSearchRepository provides a Gorm based repository
+type GormSearchRepository struct {
+	ts  *models.GormTransactionSupport
+	wir *models.GormWorkItemTypeRepository
+}
+
+// NewGormSearchRepository creates a new search repository
+func NewGormSearchRepository(ts *models.GormTransactionSupport, wir *models.GormWorkItemTypeRepository) *GormSearchRepository {
+	return &GormSearchRepository{ts, wir}
+}
+
+func generateSearchQuery(q string) (string, error) {
+	return q, nil
+}
+
+func convertFromModel(wiType models.WorkItemType, workItem models.WorkItem) (*app.WorkItem, error) {
+	result := app.WorkItem{
+		ID:      strconv.FormatUint(workItem.ID, 10),
+		Type:    workItem.Type,
+		Version: workItem.Version,
+		Fields:  map[string]interface{}{}}
+
+	for name, field := range wiType.Fields {
+		var err error
+		result.Fields[name], err = field.ConvertFromModel(name, workItem.Fields[name])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &result, nil
+}
+
+func (r *GormSearchRepository) loadTypeFromDB(ctx context.Context, name string) (*models.WorkItemType, error) {
+	log.Printf("loading work item type %s", name)
+	res := models.WorkItemType{}
+
+	if r.ts.TX().Where("name=?", name).First(&res).RecordNotFound() {
+		log.Printf("not found, res=%v", res)
+		return nil, NotFoundError{"work item type", name}
+	}
+	if err := r.ts.TX().Error; err != nil {
+		return nil, InternalError{simpleError{err.Error()}}
+	}
+
+	return &res, nil
+}
+
+//searchKeyword defines how a decomposed raw search query will look like
+type searchKeyword struct {
+	id    []string
+	words []string
+}
+
+// RegexWorkItemDetailClientURL tells us how URL for WI details on front end looks like
+const RegexWorkItemDetailClientURL = `^(?P<protocol>http[s]?)://(?P<domain>demo.almighty.io)(?P<path>/#/detail/)(?P<id>\d*)`
```

Comment:
@sbose78 sure, I am making changes: This and other pieces of code are still in progress, and in upcoming commit I will make these changes.
thanks for pointing out things.